src/coprocessor/codec/mysql/duration.rs

```diff
@@ -13,215 +13,518 @@
+    #[inline]
+    pub fn secs(self) -> u64 {
+        self.second()
     }
 
-    pub fn secs(&self) -> u64 {
-        self.dur.as_secs() % SECS_PER_MINUTE
+    /// Return the `nano` field in microseconds
+    #[inline]
+    pub fn micro_secs(self) -> u64 {
+        self.nano() / NANOS_PER_MICRO
     }
 
-    pub fn micro_secs(&self) -> u32 {
-        self.dur.subsec_micros()
+    /// Return the `nano` field
+    #[inline]
+    pub fn nano_secs(self) -> u64 {
+        self.nano()
     }
 
-    pub fn nano_secs(&self) -> u32 {
-        self.dur.subsec_nanos()
+    /// Returns the number of whole seconds contained by this Duration.
+    #[inline]
+    pub fn as_secs(self) -> u64 {
+        self.hour() * SECS_PER_HOUR + self.minute() * SECS_PER_MINUTE + self.second()
     }
 
-    pub fn to_secs(&self) -> f64 {
-        let res = self.dur.as_secs() as f64 + f64::from(self.dur.subsec_nanos()) * 10e-9;
-        if self.neg {
+    /// Return self in seconds including fractional part
+    #[inline]
+    pub fn to_secs(self) -> f64 {
+        let nonfrac = self.as_secs() as f64;
+        let frac = self.nano() as f64 / NANOS_PER_SEC as f64;
+        let res = nonfrac + frac;
+        if self.neg() {
             -res
         } else {
             res
         }
     }
 
-    pub fn is_zero(&self) -> bool {
-        self.to_nanos() == 0
+    /// Return `true` if self is zero
+    #[inline]
+    pub fn is_zero(self) -> bool {
+        self.strip() == 0
     }
 
-    pub fn to_nanos(&self) -> i64 {
-        let nanos = self.dur.as_secs() as i64 * NANOS_PER_SEC + i64::from(self.dur.subsec_nanos());
-        if self.neg {
+    /// Return `self` in nanoseconds
+    #[inline]
+    pub fn to_nanos(self) -> i64 {
+        let nanos = (self.as_secs() * NANOS_PER_SEC + self.nano()) as i64;
+        if self.neg() {
             -nanos
         } else {
             nanos
         }
     }
 
+    /// Try to build a `Duration` via nanoseconds and fsp
+    #[inline]
     pub fn from_nanos(nanos: i64, fsp: i8) -> Result<Duration> {
         let neg = nanos < 0;
         let nanos = nanos.abs();
+        Duration::new(StdDuration::from_nanos(nanos as u64), neg, fsp)
+    }
+
+    /// Try to build a `Duration` via `DurationBuilder`
+    /// NOTE:
+    /// 1. It will automatically round the `nano` field according to `fsp`
+    /// 2. `DurationBuilder.nano` can be greater than `NANOS_PER_SEC`.
+    /// 3. Return an `Err` only if the result's `hour` is invalid(> 838).
+    #[inline]
+    fn build(builder: DurationBuilder) -> Result<Duration> {
+        let DurationBuilder {
+            neg,
+            mut hour,
+            mut minute,
+            mut second,
+            mut nano,
+            fsp,
+            round_with_fsp,
+        } = builder.check()?;
+
+        if round_with_fsp {
+            let round = u64::from(TEN_POW[NANO_WIDTH as usize - fsp as usize - 1]);
+            nano /= round;
+            nano = nano / 10 + if nano % 10 > 4 { 1 } else { 0 };
```

Comment:
This is typically written as

```suggestion
            nano = (nano + 5) / 10;
```